### ISSUE TITLE
Environment and pre-defined variables are not supported in `forwardPorts` and keys of `portsAttributes`

### ISSUE THREAD

**Issue by i-ky** [devcontainers/spec#489]
To avoid hardcoding hostnames I am using environment variables, e.g. `${POSTGRES_HOSTNAME}`. This allows pointing application or any supplementary scripts to the correct host depending on the environment (e.g. in and out of dev container).
Same variables are used as `hostname:` in Docker Compose files:
```yaml
services:
  postgres:
    image: postgres:latest
    hostname: ${POSTGRES_HOSTNAME}
    ...
  ...
```
And same variables should be used in `forwardPorts` and `portsAttributes` of dev container configuration:
```json
{
    ...
    "forwardPorts": [
        "${containerEnv:POSTGRES_HOSTNAME}:5432",
        ...
    ],
    "portsAttributes": {
        "${containerEnv:POSTGRES_HOSTNAME}:5432": {
            ...
        },
        ...
    },
    ...
}
```
Unfortunately, `forwardPorts` and `portsAttributes` don't support variables, so it needs to be hardcoded instead:
```json
{
    ...
    "forwardPorts": [
        "postgres:5432",
        ...
    ],
    "portsAttributes": {
        "postgres:5432": {
            ...
        },
        ...
    },
    ...
}
```
...meaning that hostname needs to be changed manually in two more places if the environment variable changes.

**Comment by chrmarti** [comment]
Does `${localEnv:POSTGRES_HOSTNAME}` work?

**Comment by i-ky** [comment]
Hi @chrmarti!

Thank you for looking into this.

I am not sure what "work" means if we are talking about the spec here. Do you have a specific implementation in mind?

https://containers.dev/implementors/json_reference/ seems to talk about pre-defined variables as a whole. Either
> Environment and [pre-defined variables](https://containers.dev/implementors/json_reference/#variables-in-devcontainerjson) may be referenced in the values.

...or nothing at all. So my expectation is that if `containerEnv:` does not work, neither will `localEnv:`.

**Comment by chrmarti** [comment]
`containerEnv:` only works with container env variables and I don't see from your report if you set `POSTGRES_HOSTNAME` in the container. I do see that you are using the local env variable `POSTGRES_HOSTNAME` in the docker-compose.yml, so `localEnv:` might work, depending on where you set `POSTGRES_HOSTNAME` locally.

**Comment by i-ky** [comment]
> I don't see from your report if you set `POSTGRES_HOSTNAME` in the container. I do see that you are using the local env variable `POSTGRES_HOSTNAME` in the docker-compose.yml, so `localEnv:` might work, depending on where you set `POSTGRES_HOSTNAME` locally.

Sorry, I didn't mention it. `POSTGRES_HOSTNAME` is set in `.env` file (from there `POSTGRES_HOSTNAME` gets into Compose file) and `.env` file is also passed to the dev container in Compose file as `env_file:` (this is how `POSTGRES_HOSTNAME` gets into container environment).

Since `POSTGRES_HOSTNAME` is not set in host environment, I think that `localEnv:` will not work. Do you want me to try?

**Comment by chrmarti** [comment]
> Since `POSTGRES_HOSTNAME` is not set in host environment, I think that `localEnv:` will not work. Do you want me to try?

You're right. Support for env files has come up before. Related: https://github.com/microsoft/vscode-remote-release/issues/3007

**Comment by i-ky** [comment]
> > Since `POSTGRES_HOSTNAME` is not set in host environment, I think that `localEnv:` will not work. Do you want me to try?
> 
> You're right. Support for env files has come up before. Related: [microsoft/vscode-remote-release#3007](https://github.com/microsoft/vscode-remote-release/issues/3007)

Are you implying that if `POSTGRES_HOSTNAME` _was_ set in host environment then `localEnv:` _would_ work in `forwardPorts`?
